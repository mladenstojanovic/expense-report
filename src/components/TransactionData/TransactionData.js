import React from 'react';
import { useSelector } from 'react-redux';
import {
  TransactionDataSingleStyle,
  TransactionDataNumbersStyle
} from './TransactionData.style';

const TransactionData = () => {
  const transactionData = useSelector(state => state.network.transactionData);
  return Object.keys(transactionData).map(category => (
    <TransactionDataSingleStyle key={category}>
      <h1>{transactionData[category].title}</h1>
      <TransactionDataNumbersStyle>
        <p>
          Number of transactions: {transactionData[category].numOfTransactions}
        </p>
        <p>Total spent: {transactionData[category].total.toFixed(2)}</p>
        <p>
          Average amount spent:{' '}
          {(
            transactionData[category].total /
            transactionData[category].numOfTransactions
          ).toFixed(2)}
        </p>
      </TransactionDataNumbersStyle>
    </TransactionDataSingleStyle>
  ));
};

export default TransactionData;
