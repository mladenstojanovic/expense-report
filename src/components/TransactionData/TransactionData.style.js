import styled from 'styled-components/macro';

export const TransactionDataSingleStyle = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 80%;
  margin-bottom: 30px;

  h1 {
    margin-bottom: 10px;
    color: ${props => props.theme.colors.lightBlue};
  }

  p {
    margin-bottom: 3px;
  }
`;

export const TransactionDataNumbersStyle = styled.div`
  display: flex;
  flex-direction: column;
  margin-left: 30px;
`;
