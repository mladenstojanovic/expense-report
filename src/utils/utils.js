import React from 'react';
import { render } from '@testing-library/react';
import { createStore, combineReducers } from 'redux';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import networkReducer from '../store/reducers/network.reducer';
import theme from '../styles/theme';

export const sleep = milliseconds => {
  return new Promise(resolve => setTimeout(resolve, milliseconds));
};

export const sortTransactionData = data => {
  const sortedData = {};
  for (let index = 0; index < data.length; index++) {
    if (!data[index].subClass) {
      continue;
    }
    if (!sortedData[data[index].subClass.code]) {
      sortedData[data[index].subClass.code] = {
        total: 0,
        numOfTransactions: 0,
        title: data[index].subClass.title
      };
    }
    sortedData[data[index].subClass.code].total += Math.abs(
      parseInt(data[index].amount)
    );
    sortedData[data[index].subClass.code].numOfTransactions++;
  }
  return sortedData;
};

export const isEmailValid = email => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);

export const isPhoneNumberValid = phone =>
  /^[+]?[(]?[0-9]{3}[)]?[-\s.]?[0-9]{3}[-\s.]?[0-9]{4,6}$/.test(phone);

export const renderWithReduxAndStyles = (
  ui,
  {
    initialState,
    store = createStore(
      combineReducers({ network: networkReducer }),
      initialState
    )
  } = {}
) => ({
  ...render(
    <Provider store={store}>
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </Provider>
  ),
  store
});
