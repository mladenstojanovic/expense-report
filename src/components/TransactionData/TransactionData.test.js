import React from 'react';
import TransactionData from './TransactionData';
import { renderWithReduxAndStyles } from '../../utils/utils';

const initialState = {
  network: {
    transactionData: {
      '123': {
        numOfTransactions: 3,
        total: 300,
        title: 'Test1'
      },
      '321': {
        numOfTransactions: 2,
        total: 1001,
        title: 'Test2'
      }
    }
  }
};

describe('Test TransactionData component', () => {
  it('should render correctly with data in place', () => {
    const { container, getByText } = renderWithReduxAndStyles(
      <TransactionData />,
      {
        initialState
      }
    );

    const {
      network: { transactionData }
    } = initialState;

    expect(container.firstChild).toMatchSnapshot();
    expect(
      getByText(
        `Number of transactions: ${transactionData['123'].numOfTransactions}`
      )
    ).toBeInTheDocument();
    expect(
      getByText(
        `Number of transactions: ${transactionData['321'].numOfTransactions}`
      )
    ).toBeInTheDocument();
    expect(
      getByText(`Total spent: ${transactionData['123'].total.toFixed(2)}`)
    ).toBeInTheDocument();
    expect(
      getByText(`Total spent: ${transactionData['321'].total.toFixed(2)}`)
    ).toBeInTheDocument();
    expect(
      getByText(
        `Average amount spent: ${(
          transactionData['123'].total /
          transactionData['123'].numOfTransactions
        ).toFixed(2)}`
      )
    ).toBeInTheDocument();
    expect(
      getByText(
        `Average amount spent: ${(
          transactionData['321'].total /
          transactionData['321'].numOfTransactions
        ).toFixed(2)}`
      )
    ).toBeInTheDocument();
  });
});
