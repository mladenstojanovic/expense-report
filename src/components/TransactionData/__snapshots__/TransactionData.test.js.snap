// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TransactionData component should render correctly with data in place 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 80%;
  margin-bottom: 30px;
}

.c0 h1 {
  margin-top: 0;
  margin-bottom: 10px;
  color: #09d3ac;
}

.c0 p {
  margin-bottom: 3px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-left: 30px;
}

<div
  class="c0"
>
  <h1>
    Test1
  </h1>
  <div
    class="c1"
  >
    <p>
      Number of transactions: 
      3
    </p>
    <p>
      Total spent: 
      300.00
    </p>
    <p>
      Average amount spent:
       
      100.00
    </p>
  </div>
</div>
`;
